toolchain/check/import_ref.cpp

```diff
@@ -1069,6 +1056,28 @@ static auto GetLocalNameScopeId(ImportRefResolver& resolver,
                name_scope_inst);
 }
 
+// Translates a NameScopeId from the import IR to a local NameScopeId. Adds
+// unresolved constants to the resolver's work stack.
+static auto GetLocalNameScopeId(ImportRefResolver& resolver,
+                                SemIR::NameScopeId name_scope_id)
+    -> SemIR::NameScopeId {
+  // Get the instruction that created the scope.
+  auto [inst_id, inst] =
+      resolver.import_name_scopes().GetInstIfValid(name_scope_id);
+  if (!inst) {
+    // Map scopes that aren't associated with an instruction to `None`. For now,
+    // such scopes aren't used, and we don't have a good way to remap them.
+    return SemIR::NameScopeId::None;
+  }
+
+  // Get the constant value for the scope.
+  auto const_id = GetLocalConstantId(resolver, inst_id);
+  auto result = GetLocalNameScopeIdImpl(resolver, const_id);
+  CARBON_CHECK((name_scope_id.has_value() == result.has_value()) ||
+               resolver.HasNewWork());
```

Comment:
I think we can do a bit better than giving up on this check if `HasNewWork` -- whenever we succeed in getting a `const_id`, we should also succeed in getting a `name_scope_id` here:
```suggestion
  auto const_id = GetLocalConstantId(resolver, inst_id);
  if (!const_id.has_value()) {
    CARBON_CHECK(resolver.HasNewWork());
    return SemIR::NameScopeId::None;
  }
  auto result = GetLocalNameScopeIdImpl(resolver, const_id);
  CARBON_CHECK(result.has_value());
```

Reply:
How do you feel about skipping the CHECK is the `!const_id.has_value()` case? `GetLocalConstantId` basically has that invariant already (either returns non-none, or none + new work) & I tend to err on the side of not repeating things that are "too" redundant. Which is vague...

I could go either way, but figured it might be worth a little discussion.

Reply:
Yeah, I think that makes sense. (I actually initially wrote that but then added in the check because I was worried about the case where `inst_id` is somehow `None`, but I think it's pretty clear that that won't happen, because we'd have bailed out earlier because `inst` would be `nullopt` too.) Could also remove the early exit and check for `!const_id.has_value() || result.has_value()` here if you prefer. I'm happy with any of those options, feel free to pick one and merge :)

Reply:
I have a penchant for early returns & simpler expressions, so went that way in [b49eaab](https://github.com/carbon-language/carbon-lang/pull/5584/commits/b49eaab7fbafd9d437485488d8ac498ddd51b70c)